### PR TITLE
refactor: convert index to typescript

### PR DESCRIPTION
**Pull request by LiKang6688**

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -23,6 +23,16 @@ export interface ExtendedTranslator extends stardust.Translator {
   language(): string;
 }
 
+export interface Galaxy {
```
Comment (haxxmaxx) [inline src/types.ts:26]: is galaxy a term that we use for this in other places? I don't think we should introduce terms like this in the nebula eco-system unless it is thought through.

File: src/types.ts [hunk 2]
```diff
@@ -23,6 +23,16 @@ export interface ExtendedTranslator extends stardust.Translator {
...
+  translator: ExtendedTranslator;
```
Comment (haxxmaxx) [inline src/types.ts:27]: is the translator on here? I thought we got it from the useTheme hook?

File: src/types.ts [hunk 3]
```diff
@@ -101,3 +111,15 @@ export interface Column {
   stylingInfo: string[];
   sortDirection: string;
 }
+
+export interface RenderWithCarbonArguments {
+  env: Galaxy;
+  rootElement: HTMLElement;
+  model: EngineAPI.IGenericObject | undefined;
```
Comment (haxxmaxx) [inline src/types.ts:118]: Could this be optional instead of `| undefined`?

Comment (LiKang6688) [reply src/types.ts:118]: Yes

File: src/types.ts (hunk 1 above)
Comment (LiKang6688) [reply src/types.ts:26]: Yes, https://qlik.dev/libraries-and-tools/nebulajs/in-qlik-sense#exposed-from-qlik-sense and https://github.com/qlik-oss/sn-pivot-table/blob/main/src/index.ts#L15

File: src/types.ts (hunk 2 above)
Comment (LiKang6688) [reply src/types.ts:27]: Yes, the translator is from Qlik Sense https://qlik.dev/libraries-and-tools/nebulajs/in-qlik-sense

File: src/index.ts [hunk 4]
```diff
@@ -62,12 +73,14 @@ export default function supernova(env) {
...
-      }, [layout, pageInfo]);
+      const [tableData] = usePromise(
+        async () => (env.carbon ? nothing() : manageData(model, layout, pageInfo, setPageInfo)),
+        [layout, pageInfo]
+      );
 
       useEffect(() => {
-        if (!env.carbon && reactRoot && layout && tableData && announce && changeSortOrder && theme) {
+        const isReadyToRender = !env.carbon && reactRoot && layout && tableData && announce && changeSortOrder && theme;
+        isReadyToRender &&
```
Comment (haxxmaxx) [inline src/index.ts:83]: was this to fix the problem with too many things in the if condition?

File: src/types.ts (hunk 1 above)
Comment (haxxmaxx) [reply src/types.ts:26]: great

File: src/index.ts (hunk 4 above)
Comment (LiKang6688) [reply src/index.ts:83]: yes

Comment (LiKang6688) [reply src/index.ts:83]: codeclimate complains

File: src/types.ts [hunk 5]
```diff
@@ -61,7 +71,7 @@ export interface PaginationColors {
 }
 export interface TableThemeColors {
   tableBackgroundColorFromTheme: string;
-  backgroundColor: string | undefined;
+  backgroundColor?: string;
```
Comment (haxxmaxx) [inline src/types.ts:74]: Are you sure this is the same thing? This is set to undefined sometimes, not unset. It might be the same thing but I am not sure

**Review by haxxmaxx — APPROVED** [review]

File: src/types.ts (hunk 5 above)
Comment (LiKang6688) [reply src/types.ts:74]: undefined happens when we don’t assign a value to a variable, so unset === set to undefined

Comment (haxxmaxx) [reply src/types.ts:74]: ok

**Pull request merged**